components/job-orchestration/job_orchestration/scheduler/search/search_scheduler.py

```diff
@@ -167,51 +176,77 @@ def dispatch_search_job(
     job.current_sub_job_async_task_result = task_group.apply_async()
 
 
-def handle_pending_search_jobs(
-    db_conn, results_cache_uri: str, num_archives_to_search_per_sub_job: int
+async def handle_pending_search_jobs(
+    db_conn, results_cache_uri: str, num_archives_to_search_per_sub_job: int, jobs_poll_delay: float
 ) -> None:
     global active_jobs
-    pending_jobs = [
-        job for job in active_jobs.values() if job.current_sub_job_async_task_result is None
-    ]
-    with contextlib.closing(db_conn.cursor(dictionary=True)) as cursor:
-        for job in fetch_new_search_jobs(cursor):
-            job_id = str(job["job_id"])
-            search_config = SearchConfig.parse_obj(msgpack.unpackb(job["search_config"]))
-            archives_for_search = get_archives_for_search(db_conn, search_config)
-            if len(archives_for_search) == 0:
-                if set_job_status(
-                    db_conn, job_id, SearchJobStatus.SUCCEEDED, SearchJobStatus.PENDING
-                ):
-                    logger.info(f"No matching archives, skipping job {job['job_id']}.")
-                continue
-            new_search_job = SearchJob(
-                id=job_id,
-                search_config=search_config,
-                remaining_archives_for_search=archives_for_search,
-                current_sub_job_async_task_result=None,
-            )
-            active_jobs[job_id] = new_search_job
-            pending_jobs.append(new_search_job)
-        db_conn.commit()
 
-    for job in pending_jobs:
-        job_id = job.id
+    while True:
+        pending_jobs = [
+            job for job in active_jobs.values() if job.current_sub_job_async_task_result is None
+        ]
+        with contextlib.closing(db_conn.cursor(dictionary=True)) as cursor:
+            for job in fetch_new_search_jobs(cursor):
+                job_id = str(job["job_id"])
+                search_config = SearchConfig.parse_obj(msgpack.unpackb(job["search_config"]))
+                archives_for_search = get_archives_for_search(db_conn, search_config)
+                if len(archives_for_search) == 0:
+                    if set_job_status(
+                        db_conn, job_id, SearchJobStatus.SUCCEEDED, SearchJobStatus.PENDING
+                    ):
+                        logger.info(f"No matching archives, skipping job {job['job_id']}.")
+                    continue
+
+                reducer_recv_handle = None
+                reducer_send_handle = None
+                if search_config.count is not None:
+                    search_config.job_id = job["job_id"]
+                    # Get a reducer and wait for it to be ready
+                    while True:
+                        reducer_host, reducer_port, reducer_recv_handle, reducer_send_handle = (
+                            await reducer_connection_queue.get()
```

Comment:
If I'm reading the code correctly, there's a potential race condition between the two "threads" which handle pending and cancelling jobs:

* Let's say we get a pending job from the DB, get a reducer for it, add it to `active_jobs`, and add it to `pending_jobs`.
* Then for the next pending job, when we call `await` to wait for a reducer, we may context switch to `handle_cancelling_search_jobs`.
* At this point, the user might've cancelled the search job, meaning we'll remove it from active_jobs.
* When we eventually context switch back to `handle_pending_search_jobs` and move on to the `for job in pending_jobs` loop, we'll operate on a job that's actually been cancelled already.

In the bigger picture, to use asyncio with all of this code, (even if my analysis above is wrong) it feels like we need clearer boundaries about where we can and cannot modify state that's shared between the "threads".

Reply:
Gotcha.

Thinking about it a bit more, one potential issue is that if we have several pending aggregation queries, each of which only needs a few workers, we may exhaust the number of reducers in the cluster before we exhaust the number of workers. That means some pending non-aggregation queries may be blocked because a previous aggregation query is waiting for a reducer. Similarly, if a user cancels one of these blocked aggregation queries, I think the cancellation would hang because it's waiting for a reducer, right?

So in that sense, I feel like we should move the reducer acquisition and connection negotiation out of the pending jobs loop. This would not only avoid blocking the pending jobs loop but it would make it easier to reason about potential race conditions since collecting and starting pending jobs would become atomic again. To keep track of the job's state, I think we'd need to maintain an internal state class (differentiated from the job state we store in MySQL), and one of the states could be `WAITING_FOR_REDUCER` (like we had before).

Now the remaining (slightly complicated) part would be how to assign reducers to active jobs without blocking processing pending/cancelling jobs, and without creating race conditions. I think we could create an asyncio task per aggregation job and store the task both in the `SearchJob` class as well the tasks we wait for in the `handle_jobs` loop. When one of these pending tasks completes, we change the job's status to `WAITING_FOR_ARCHIVES` and let `handle_pending_jobs` take care of it. If an aggregation job is cancelled before it gets a reducer, we would cancel the asyncio task and handle the cancelled error in `handle_jobs`.

What do you think?

Reply:
I think this generally makes sense to me. We might have to be a bit careful about cleaning up state after cancelling reducer acquisition to avoid hanging reducers, but that should be do-able. I'm a bit concerned about how moving reducer acquistion into an async task will add a polling interval worth of start-up time to every aggregation task, but it shouldn't be too hard to start the celery tasks from the handle_jobs loop as soon as the reducer is available if it's a real issue.

Reply:
I think one potential solution to the polling interval issue is:
* Don't loop in `handle_pending_search_jobs`; instead return the new asyncio tasks directly and let `handle_jobs` be the one to loop.
* *Only* sleep for the poll delay if there were *no* pending jobs